src/bin/hive_chain.rs

```diff
@@ -0,0 +1,78 @@
+use alloy_rlp::Decodable;
+use eyre::eyre;
+use kakarot_rpc::eth_provider::starknet::kakarot_core::to_starknet_transaction;
+use reth_primitives::{bytes::Buf, Block, BlockBody, BytesMut};
+use starknet::{
+    core::types::BroadcastedInvokeTransaction,
+    providers::{jsonrpc::HttpTransport, JsonRpcClient, Provider},
+};
+use starknet_crypto::FieldElement;
+use std::{path::Path, str::FromStr};
+use tokio::{fs::File, io::AsyncReadExt};
+use tokio_stream::StreamExt;
+use tokio_util::codec::{Decoder, FramedRead};
+use url::Url;
+
+struct BlockFileCodec;
+
+impl Decoder for BlockFileCodec {
+    type Item = Block;
+    type Error = eyre::Error;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        if src.is_empty() {
+            return Ok(None);
+        }
+        let buf_slice = &mut src.as_ref();
+        let body = Block::decode(buf_slice)?;
+        src.advance(src.len() - buf_slice.len());
+        Ok(Some(body))
+    }
+}
+
+/// Inspired by the Import command from Reth.
+/// https://github.com/paradigmxyz/reth/blob/main/bin/reth/src/commands/import.rs
+#[tokio::main]
+async fn main() -> eyre::Result<()> {
+    let chain_path = Path::new(&std::env::var("CHAIN_PATH")?).to_path_buf();
+    let mut file = File::open(chain_path).await?;
+
+    let metadata = file.metadata().await?;
+    let file_len = metadata.len();
+
+    // read the entire file into memory
+    let mut reader = vec![];
+    file.read_to_end(&mut reader).await.unwrap();
+    let mut stream = FramedRead::with_capacity(&reader[..], BlockFileCodec, file_len as usize);
+
+    let mut bodies = Vec::new();
+    while let Some(block_res) = stream.next().await {
+        let block = block_res?;
+
+        bodies.push(BlockBody { transactions: block.body, ommers: block.ommers, withdrawals: block.withdrawals });
+    }
+
+    let provider = JsonRpcClient::new(HttpTransport::new(Url::from_str(&std::env::var("STARKNET_NETWORK")?)?));
+    let mut current_nonce = FieldElement::ZERO;
+
+    for (block_number, body) in bodies.into_iter().enumerate() {
+        while provider.block_number().await? < block_number as u64 {
+            tokio::time::sleep(tokio::time::Duration::from_millis(500)).await;
```

Comment:
why are we sleeping 500ms here

Reply:
arbitrary value, just to wait for the katana block to finish

Reply:
but block is 2000ms no?
should we use wait for tx? or smth?

Reply:
my fear is that this might not work in every run ~ have UB

Reply:
I'm not sure what you mean by UB, you can't really have UB unless the block time is lower than 500ms.

Reply:
you said wait for the katana block to finish, but with 500ms of sleep, there can be up to 3 txs in a single block of katana with 2000ms ?

Reply:
No so what this does is:
- Current block is 0 from chain rlp and Katana is at 0 -> send all the txs for this block.
- Wait for the Katana block to be 1, querying every 500 ms.
- Send all the txs for block 1.
- ...